lsp: text_document.version may be vim.NIL not nil

h-michael opened the pull request:
I take over #12191

cc: @ckipp01

ckipp01 committed:
[LSP] check for vim.NIL and add apply_text_document_edit tests

h-michael committed:
lsp: fix apply_text_document_edit test

lsp.util.buf_versions must be set in advance.
Use helper.insert to create an anonymous buffer,
so create a named buffer for testing without using insert.

ckipp01 commented on test/functional/plugin/lsp_spec.lua:
```diff
@@ -817,6 +817,87 @@ describe('LSP', function()
+  describe('apply_text_document_edit', function()
+    local target_bufnr
+    before_each(function()
+      target_bufnr = exec_lua [[
+        local bufnr = vim.fn.bufadd("fake/uri")
+        local lines = {"1st line of text", "2nd line of text"}
+        vim.api.nvim_buf_set_lines(bufnr, 0, 1, false, lines)
+        return bufnr
+      ]]
```
Ahh, this is super helpful to see how to do. 👍

clason commented on test/functional/plugin/lsp_spec.lua:
```diff
@@ -817,6 +817,87 @@ describe('LSP', function()
+    before_each(function()
+      target_bufnr = exec_lua [[
+        local bufnr = vim.fn.bufadd("fake/uri")
+        local lines = {"1st line of text", "2nd line of text"}
```
Please add a non-ASCII character to one of the lines (maybe "Ä 2nd line of text"?) to test for encoding.

teto replied:
don't. Let's keep meaningful commits, non-ASCII characters should be merged with the same commits that introduce supports for them i.e., https://github.com/neovim/neovim/pull/12223

clason replied:
No, that commit didn't introduce support, non-ASCII characters should have worked from the start, so it fixes a bug this test would have caught.

clason replied:
(But once the test is merged, I can try to extend it with non-ASCII characters. I'd prefer not to merge this PR into my PR, though.)

clason commented on test/functional/plugin/lsp_spec.lua:
```diff
@@ -817,6 +817,87 @@ describe('LSP', function()
+        '2nd line of text';
+      }, buf_lines(target_bufnr))
+    end)
+    it('correctly goes ahead with the edit whe the version is vim.NIL', function()
```
typo: "when" (or "correctly applies the edit if the version...")

h-michael committed:
lsp: add a lsp.util.apply_text_edits test(pending)

We don't handle non-ASCII characters well in UTF-16.
So I add a non-ASCII characters test case.

h-michael commented:
@teto @clason 
I've fixed a bit and add a non-ASCII characters test case pending.
